src/Workspaces/Core/Portable/Options/SerializableOptionSet.cs

```diff
@@ -133,8 +133,16 @@ private bool ShouldSerialize(OptionKey optionKey)
 
         public override OptionSet WithChangedOption(OptionKey optionKey, object? value)
         {
-            // make sure we first load this in current optionset
-            this.GetOption(optionKey);
+            // Make sure we first load this in current optionset
+            var currentValue = this.GetOption(optionKey);
+
+            // Check if the new value is the same as the current value.
+            if (value == currentValue)
+            {
+                // Return a cloned option set as the public API 'WithChangedOption' guarantees a new option set is returned.
+                return new SerializableOptionSet(_languages, _workspaceOptionSet, _serializableOptions,
+                    _serializableOptionValues, _changedOptionKeysSerializable, _changedOptionKeysNonSerializable);
```

Comment:
> _changedOptionKeysSerializable

💭 It would be much easier to understand states if we made this field always relative to the _default_ values, instead of making them relevant to the current globals.

Reply:
Yep, that can be a separate refactoring if required.